aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu

```diff
@@ -0,0 +1,472 @@
+#include "ATen/ATen.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/TensorUtils.h"
+#include "ATen/Utils.h"
+#include "c10/util/Exception.h"
+#include <THC/THCGeneral.h>
+#include "THC/THCNumerics.cuh"
+
+#include <algorithm>
+#include <cfloat>
+#include <cmath>
+
+#define CUDA_MAX_THREADS 1024   // this is safe, in reality 256 is our limit
+
+#define START_IND(a,b,c) (int)std::floor((float)(a * c) / b)
+#define END_IND(a,b,c) (int)std::ceil((float)((a + 1) * c) / b)
+// #define START_IND(a,b,c) a * c / b
+// #define END_IND(a,b,c)  (a + 1) * c / b + ((a + 1) * c % b > 0)?1:0
+
+
+namespace at {
+namespace native {
+
+namespace {
+
+// 4d tensor B x D x H x W
+
+/*
+ * Description:
+ *    this function adaptively maxpools an input 4D tensor along dimensions 2 and 3
+ *    4D input, 4D output, 4D argmax x and y
+ */
+ template <typename T>
+__global__ void adaptivemaxpool(T *input, T *output, int64_t *indices,
+                        int isizeH, int isizeW,
+                        int osizeH, int osizeW,
+                        int64_t istrideD, int64_t istrideH, int64_t istrideW)
+{
+  // iterators
+  int oh, ow;
+
+  // compute offsets based on thread/block ID
+  int o_plane = blockIdx.x;
+  int i_plane = o_plane;
+
+  int ostartW = threadIdx.x;
+  int oendW = osizeW;
+  const int ostepW = blockDim.x;
+
+  int ostartH = blockDim.y*blockIdx.y + threadIdx.y;
+  int oendH = osizeH;
+  const int ostepH = blockDim.y*gridDim.y;
+  // select input/output plane
+  output = output + o_plane*osizeH*osizeW;
+  input = input + i_plane*istrideD;
+  indices = indices + o_plane*osizeH*osizeW;
+
+  // For all output pixels...
+  for(oh = ostartH; oh < oendH; oh += ostepH) {
+
+    int istartH = START_IND(oh, osizeH, isizeH);
+    int iendH   = END_IND(oh, osizeH, isizeH);
+    int kH = iendH - istartH;
+
+    for(ow = ostartW; ow < oendW; ow += ostepW) {
+      int istartW = START_IND(ow, osizeW, isizeW);
+      int iendW   = END_IND(ow, osizeW, isizeW);
+
+      int kW = iendW - istartW;
+
+      // Compute the mean of the input image...
+      T *ptr_input = input + istartH*istrideH + istartW*istrideW;
+      T *ptr_output = output + oh*osizeW + ow;
+      int64_t *ptr_ind = indices + oh*osizeW + ow;
+      int argmax = -1;
+      T max = THCNumerics<T>::min();
```

Comment:
In the cuda files I found references to `THCNumerics` and also to `at::numeric_limits` ... not sure if here it could be changed to `at::numeric_limits<T>::min()`

Reply:
We should... stop using THCNumerics. But there's a lot of use sites already in the code, so you don't have to fix it.